src/Integration/MefServices/ProjectsRuleSetProvider.cs

```diff
@@ -0,0 +1,355 @@
+        {
+        }
+
+        internal ProjectsRuleSetProvider(IProjectSystemHelper projectSystemHelper, IVsFileChangeEx fileChangeService,
+            IActiveSolutionTracker activeSolutionTracker, IConfigurationProvider configurationProvider, ILogger logger)
+        {
+            if (projectSystemHelper == null)
+            {
+                throw new ArgumentNullException(nameof(projectSystemHelper));
+            }
+
+            if (fileChangeService == null)
+            {
+                throw new ArgumentNullException(nameof(fileChangeService));
+            }
+
+            if (activeSolutionTracker == null)
+            {
+                throw new ArgumentNullException(nameof(activeSolutionTracker));
+            }
+
+            if (configurationProvider == null)
+            {
+                throw new ArgumentNullException(nameof(configurationProvider));
+            }
+
+            if (logger == null)
+            {
+                throw new ArgumentNullException(nameof(logger));
+            }
+
+            this.projectSystemHelper = projectSystemHelper;
+            this.fileChangeService = fileChangeService;
+            this.activeSolutionTracker = activeSolutionTracker;
+            this.configurationProvider = configurationProvider;
+            this.logger = logger;
+
+            this.activeSolutionTracker.ActiveSolutionChanged += OnActiveSolutionChanged;
+            this.activeSolutionTracker.AfterProjectOpened += OnAfterProjectOpened;
+
+            BuildCacheAsync();
+        }
+
+        private async void OnAfterProjectOpened(object sender, ProjectOpenedEventArgs e)
+        {
+            if (this.projectSystemHelper.IsSolutionFullyOpened() &&
+                e.Project != null &&
+                !string.IsNullOrEmpty(e.Project.FullName) &&
+                !projectPathToCachedData.ContainsKey(e.Project.FullName))
+            {
+                // Improve me: there is no need to rebuild the full cache here
+                ClearCache();
+                await BuildCacheAsync();
+            }
+        }
+
+        private async void OnActiveSolutionChanged(object sender, ActiveSolutionChangedEventArgs e)
+        {
+            ClearCache();
+
+            if (e.IsSolutionOpen)
+            {
+                await BuildCacheAsync();
+            }
+        }
+
+        public bool HasRuleSetWithSonarAnalyzerRules(string projectFilePath) =>
+            projectPathToCachedData.ContainsKey(projectFilePath) && projectPathToCachedData[projectFilePath].HasAnySonarRule;
+
+        private void ClearCache()
+        {
+            filesToTrack.Values.ToList().ForEach(x => x.UpdatedOnDisk -= OnUpdatedOnDisk);
+            filesToTrack.Clear();
+
+            projectPathToCachedData.Clear();
+        }
+
+        private async Task BuildCacheAsync()
+        {
+            while (!projectSystemHelper.IsSolutionFullyOpened())
```

Comment:
Is this _while_ needed? I don't think the active solution tracker will raise the event until the solution has finished loading, so it shouldn't be needed when responding to those events.
That leaves the initial construction; if there isn't a solution fully-open at the time this class is created, then couldn't we rely on the event from the active solution tracker in that case too?

Reply:
On all my tests the solution was never fully loaded before the ctor was called. The active solution tracker event was also not always raising at the expected time hence this wait.